src/main/java/org/jenkinsci/plugins/docker/workflow/WithContainerStep.java

```diff
@@ -171,16 +171,27 @@ private static FilePath tempDir(FilePath ws) {
         private static final long serialVersionUID = 1;
         private final String container;
         private final String[] envHost;
+        private final String ws;
 
-        Decorator(String container, EnvVars envHost) {
+        Decorator(String container, EnvVars envHost, String ws) {
             this.container = container;
             this.envHost = Util.mapToEnv(envHost);
+            this.ws = ws;
         }
 
-        @Override public Launcher decorate(Launcher launcher, Node node) {
+        @Override public Launcher decorate(final Launcher launcher, Node node) {
             return new Launcher.DecoratedLauncher(launcher) {
                 @Override public Proc launch(Launcher.ProcStarter starter) throws IOException {
                     List<String> prefix = new ArrayList<String>(Arrays.asList("docker", "exec", container, "env"));
+                    if (ws != null) {
+                        FilePath cwd = starter.pwd();
+                        if (cwd != null) {
+                            String path = cwd.getRemote();
+                            if (!path.equals(ws)) {
+                                launcher.getListener().getLogger().println("JENKINS-33510: working directory will be " + ws + " not " + path);
```

Comment:
hack-ish command line black magic vs dedicated binary we can silently inject into container ...